llvm/lib/Target/Mips/MipsSEISelDAGToDAG.cpp

```diff
@@ -1465,6 +1465,17 @@ SelectInlineAsmMemoryOperand(const SDValue &Op, unsigned ConstraintID,
     llvm_unreachable("Unexpected asm memory constraint");
   // All memory constraints can at least accept raw pointers.
   case InlineAsm::Constraint_m:
+    if (Subtarget->hasNanoMips()) {
+      // On NanoMips, they can only handle 9-bit offsets. Also, 2
+      // least-significant bits need to be cleared as required by LL and SC
+      // instructions.
+      if (selectAddrRegImm9(Op, Base, Offset)) {
```

Comment:
This will verify that the offset is 4-byte aligned? Better move this to Constraint_o. Also on gcc m constraint doesn't allow offsets. While i disagree with that decision it is better to follow gcc here.

Reply:
I agree the current solution isn't enough to ensure a 4-byte alignment. I will make sure to fix this.

Regarding the "o" constraint, it seems that to follow GCC logic, we may need to use something like this:
```
if (selectAddrFrameIndexOffset(Op, Base, Offset, 13)) {
        OutOps.push_back(Base);
        OutOps.push_back(Offset);
        return false;
}
```
Also, with larger offsets, GCC separates the constant in the following manner.
For example, if we have a constant 4098, GCC generates the code as follows:
```
        addiu	$a0,$a0,4096
	ll      $zero, 2($a0) 
```

Should we follow this convention, or 
```
        addiu	$a0,$a0,4098
	ll      $zero, 0($a0) 
```

would be sufficient?